src/main/java/com/timgroup/statsd/UnixStreamClientChannel.java

```diff
@@ -0,0 +1,165 @@
+            }
+        } catch (IOException e) {
+            disconnect();
+            throw e;
+        }
+
+        return size;
+    }
+
+    /**
+     * Writes all bytes from the given buffer to the channel.
+     * @param bb buffer to write
+     * @param canReturnOnTimeout if true, we return if the channel is blocking and we haven't written anything yet
+     * @param deadline deadline for the write
+     * @return number of bytes written
+     * @throws IOException if the channel is closed or an error occurs
+     */
+    public int writeAll(ByteBuffer bb, boolean canReturnOnTimeout, long deadline) throws IOException {
+        int remaining = bb.remaining();
+        int written = 0;
+        while (remaining > 0) {
+            int read = delegate.write(bb);
+
+            // If we haven't written anything yet, we can still return
+            if (read == 0 && canReturnOnTimeout && written == 0) {
+                return written;
+            }
+
+            remaining -= read;
+            written += read;
+
+            if (deadline > 0 && System.currentTimeMillis() > deadline) {
+                throw new IOException("Write timed out");
+            }
+        }
+        return written;
+    }
+
+    private void connectIfNeeded() throws IOException {
+        if (delegate == null) {
+            connect();
+        }
+    }
+
+    private void disconnect() throws IOException {
+        if (delegate != null) {
+            delegate.close();
+            delegate = null;
+        }
+    }
+
+    private void connect() throws IOException {
+        if (this.delegate != null) {
+            try {
+                disconnect();
+            } catch (IOException e) {
+                // ignore to be sure we don't stay with a broken delegate forever.
+            }
+        }
+
+        UnixSocketChannel delegate = UnixSocketChannel.create();
+
+        long deadline = System.currentTimeMillis() + connectionTimeout;
+        if (connectionTimeout > 0) {
+            // Set connect timeout, this should work at least on linux
+            // https://elixir.bootlin.com/linux/v5.7.4/source/net/unix/af_unix.c#L1696
+            // We'd have better timeout support if we used Java 16's native Unix domain socket support (JEP 380)
+            delegate.setOption(UnixSocketOptions.SO_SNDTIMEO, connectionTimeout);
+        }
+        delegate.connect(address);
+        while (!delegate.finishConnect()) {
+            // wait for connection to be established
+            try {
+                Thread.sleep(10);
+            } catch (InterruptedException e) {
+                throw new IOException("Interrupted while waiting for connection", e);
+            }
+            if (connectionTimeout > 0 && System.currentTimeMillis() > deadline) {
+                throw new IOException("Connection timed out");
+            }
+        }
+
+        if (timeout > 0) {
+            delegate.setOption(UnixSocketOptions.SO_SNDTIMEO, timeout);
+        }
```

Comment:
When using connection timeout, if this timeout is zero, will the socket will continue using connection timeout for writes?

Reply:
Good catch, setting to 0 in this case